src/high-level/matrix.lisp

```diff
@@ -0,0 +1,536 @@
+          (setf (tref result r c) (tref a r c))))
+      (loop :for r :from arows :below rrows :do
+        (loop :for c :from acols :below rcols :do
+          (setf (tref result r c) (tref b (cl:- r arows) (cl:- c acols)))))
+      result)))
+
+(defgeneric kron (a b &rest rest)
+  (:documentation "Compute the kronecker product of A and B")
+  (:method (a b &rest rest)
+    (let ((ma (nrows a))
+          (mb (nrows b))
+          (na (ncols a))
+          (nb (ncols b)))
+      (flet ((calc-i-j (i j) (* (tref a (floor i mb) (floor j nb))
+                                (tref b (mod i mb) (mod j nb)))))
+        (reduce #'kron rest :initial-value (into! #'calc-i-j
+                                                  (empty (list (* ma mb) (* na nb))
+                                                         :type '(complex double-float))))))))
+
+(defgeneric transpose! (matrix &key fast)
+  (:documentation "Transpose MATRIX, replacing the elements of MATRIX, optionally performing a faster change of order if FAST is specified")
+  (:method ((matrix matrix) &key fast)
+    "Transpose a matrix by copying values.
+If fast is t then just change order. Fast can cause problems when you want to multiply specifying transpose."
+    (if fast
+        (progn (rotatef (matrix-ncols matrix) (matrix-nrows matrix))
+               (setf (matrix-order matrix) (ecase (matrix-order matrix)
+                                             (:row-major :column-major)
+                                             (:column-major :row-major))))
+        (let ((index-function
+                (ecase (matrix-order matrix)
+                  (:row-major #'row-major-index)
+                  (:column-major #'column-major-index)))
+              (shape (shape matrix)))
+          (loop :for row :below (matrix-nrows matrix)
+                :do (loop :for col :from row :below (matrix-ncols matrix)
+                          :do (rotatef
+                               (aref (storage matrix) (funcall index-function (list row col) shape))
+                               (aref (storage matrix) (funcall index-function (list col row) shape)))))
+          (rotatef (matrix-ncols matrix) (matrix-nrows matrix))))
+    matrix))
+
+(defgeneric transpose (matrix)
+  (:documentation "Create a new matrix containing the transpose of MATRIX")
+  (:method ((matrix matrix))
+    "Transpose a matrix by copying values.
+If fast is t then just change order. Fast can cause problems when you want to multiply specifying transpose."
+    (let ((new-matrix (copy-tensor matrix)))
+      (let ((index-function
+              (ecase (order matrix)
+                (:row-major #'row-major-index)
+                (:column-major #'column-major-index)))
+            (shape (shape matrix)))
+        (loop :for row :below (matrix-nrows matrix)
+              :do(loop :for col :from row :below (matrix-ncols matrix)
+                       :do (let ((index1 (funcall index-function (list row col) shape))
```

Comment:
This seems to fail under row major indexing. For example, `(magicl:transpose (magicl:from-list '(0d0 1d0) '(1 2) :order :row-major))` gives the following error

`Invalid index 2 for (SIMPLE-ARRAY DOUBLE-FLOAT (2)), should be a non-negative integer below 2.Invalid index 2 for (SIMPLE-ARRAY DOUBLE-FLOAT (2)), should be a non-negative integer below 2.`

Reply:
Fixed in 1328947